File: migration/nocodb_client.py
"""
NocoDB REST API Client
NocoDBのREST APIラッパー（プロジェクト、テーブル、フィールド、レコード操作）
"""

import requests
import logging
from typing import Dict, List, Optional, Any
from time import sleep

logger = logging.getLogger(__name__)


class NocoDBClient:
    """NocoDB REST API クライアント"""

    def __init__(self, base_url: str, api_token: str):
        """
        Args:
            base_url: NocoDB URL (例: http://localhost:8080)
            api_token: NocoDB API Token
        """
        self.base_url = base_url.rstrip('/')
        self.api_token = api_token
        self.headers = {
            'xc-token': api_token,
            'Content-Type': 'application/json'
        }

    def _request(self, method: str, endpoint: str, **kwargs) -> requests.Response:
        """
        HTTP リクエストを実行（リトライ機構付き）

        Args:
            method: HTTPメソッド (GET, POST, PUT, DELETE)
            endpoint: APIエンドポイント
            **kwargs: requestsライブラリに渡す引数

        Returns:
            requests.Response
        """
        url = f"{self.base_url}/api/v1{endpoint}"
        max_retries = 3
        retry_delay = 2

        for attempt in range(max_retries):
            try:
                response = requests.request(
                    method=method,
                    url=url,
                    headers=self.headers,
                    **kwargs
                )
                response.raise_for_status()
                return response
            except requests.exceptions.RequestException as e:
                logger.warning(f"Request failed (attempt {attempt + 1}/{max_retries}): {e}")
                # エラーレスポンスの詳細を出力
                if hasattr(e, 'response') and e.response is not None:
                    try:
                        error_detail = e.response.json()
                        logger.error(f"Error response: {error_detail}")
                    except:
                        logger.error(f"Error response (text): {e.response.text[:500]}")

                if attempt < max_retries - 1:
                    sleep(retry_delay * (attempt + 1))
                else:
                    raise

    # === プロジェクト操作 ===

    def create_project(self, title: str, description: str = "") -> Dict[str, Any]:
        """
        新規プロジェクト作成

        Args:
            title: プロジェクト名
            description: 説明

        Returns:
            プロジェクト情報
        """
        logger.info(f"Creating project: {title}")
        response = self._request(
            'POST',
            '/db/meta/projects',
            json={
                'title': title,
                'description': description
            }
        )
        return response.json()

    def list_projects(self) -> List[Dict[str, Any]]:
        """
        プロジェクト一覧取得

        Returns:
            プロジェクトリスト
        """
        response = self._request('GET', '/db/meta/projects')
        return response.json().get('list', [])

    def get_project(self, project_id: str) -> Dict[str, Any]:
        """
        プロジェクト詳細取得

        Args:
            project_id: プロジェクトID

        Returns:
            プロジェクト情報
        """
        response = self._request('GET', f'/db/meta/projects/{project_id}')
        return response.json()

    # === テーブル操作 ===

    def create_table(self, project_id: str, title: str, columns: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        新規テーブル作成

        Args:
            project_id: プロジェクトID
            title: テーブル名
            columns: カラム定義リスト

        Returns:
            テーブル情報
        """
        logger.info(f"Creating table: {title}")
        response = self._request(
            'POST',
            f'/db/meta/projects/{project_id}/tables',
            json={
                'table_name': title,
                'title': title,
                'columns': columns
            }
        )
        return response.json()

    def list_tables(self, project_id: str) -> List[Dict[str, Any]]:
        """
        テーブル一覧取得

        Args:
            project_id: プロジェクトID

        Returns:
            テーブルリスト
        """
        response = self._request('GET', f'/db/meta/projects/{project_id}/tables')
        return response.json().get('list', [])

    def get_table(self, table_id: str) -> Dict[str, Any]:
        """
        テーブル詳細取得

        Args:
            table_id: テーブルID

        Returns:
            テーブル情報
        """
        response = self._request('GET', f'/db/meta/tables/{table_id}')
        return response.json()

    # === カラム操作 ===

    def create_column(self, table_id: str, column: Dict[str, Any]) -> Dict[str, Any]:
        """
        新規カラム作成

        Args:
            table_id: テーブルID
            column: カラム定義

        Returns:
            カラム情報
        """
        logger.info(f"Creating column: {column.get('column_name')}")
        response = self._request(
            'POST',
            f'/db/meta/tables/{table_id}/columns',
            json=column
        )
        return response.json()

    def list_columns(self, table_id: str) -> List[Dict[str, Any]]:
        """
        カラム一覧取得

        Args:
            table_id: テーブルID

        Returns:
            カラムリスト
        """
        response = self._request('GET', f'/db/meta/tables/{table_id}/columns')
        return response.json().get('list', [])

    # === レコード操作 ===

    def create_record(self, project_id: str, table_name: str, record: Dict[str, Any]) -> Dict[str, Any]:
        """
        新規レコード作成

        Args:
            project_id: プロジェクトID
            table_name: テーブル名
            record: レコードデータ

        Returns:
            作成されたレコード
        """
        response = self._request(
            'POST',
            f'/db/data/noco/{project_id}/{table_name}',
            json=record
        )
        return response.json()

    def bulk_create_records(self, project_id: str, table_name: str, records: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        一括レコード作成

        Args:
            project_id: プロジェクトID
            table_name: テーブル名
            records: レコードデータリスト

        Returns:
            作成されたレコードリスト
        """
        logger.info(f"Bulk creating {len(records)} records in {table_name}")
        response = self._request(
            'POST',
            f'/db/data/bulk/noco/{project_id}/{table_name}',
            json=records
        )
        return response.json()

    def list_records(
        self,
        project_id: str,
        table_name: str,
        limit: int = 100,
        offset: int = 0,
        where: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        レコード一覧取得

        Args:
            project_id: プロジェクトID
            table_name: テーブル名
            limit: 取得件数
            offset: オフセット
            where: フィルタ条件

        Returns:
            レコードリストとページング情報
        """
        params = {
            'limit': limit,
            'offset': offset
        }
        if where:
            params['where'] = where

        response = self._request(
            'GET',
            f'/db/data/noco/{project_id}/{table_name}',
            params=params
        )
        return response.json()

    def update_record(
        self,
        project_id: str,
        table_name: str,
        record_id: str,
        data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        レコード更新

        Args:
            project_id: プロジェクトID
            table_name: テーブル名
            record_id: レコードID
            data: 更新データ

        Returns:
            更新されたレコード
        """
        response = self._request(
            'PATCH',
            f'/db/data/noco/{project_id}/{table_name}/{record_id}',
            json=data
        )
        return response.json()

    def delete_record(self, project_id: str, table_name: str, record_id: str) -> None:
        """
        レコード削除

        Args:
            project_id: プロジェクトID
            table_name: テーブル名
            record_id: レコードID
        """
        self._request(
            'DELETE',
            f'/db/data/noco/{project_id}/{table_name}/{record_id}'
        )

File: mcp/nocodb/src/nocodb-client.ts
/**
 * NocoDB REST API Client (Airtable-compatible interface)
 * Airtable APIと同じインターフェースを提供し、既存コードの移行を簡易化
 */

import axios, { AxiosInstance } from 'axios';
import { readFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export interface ListOptions {
  where?: string;
  limit?: number;
  offset?: number;
  sort?: string;
  fields?: string[];
}

export interface ColumnOption {
  title: string;
  color?: string;
}

export interface ColumnUpdateOptions {
  colOptions?: {
    options: ColumnOption[];
  };
}

export interface TableMeta {
  id: string;
  title: string;
  columns: ColumnMeta[];
}

export interface ColumnMeta {
  id: string;
  title: string;
  uidt: string;
  dtxp?: string;
  colOptions?: {
    options: Array<{
      id: string;
      title: string;
      color?: string;
    }>;
  };
}

export interface AirtableRecord {
  id: string;
  fields: Record<string, any>;
  createdTime: string;
}

export class NocoDBClient {
  private baseUrl: string;
  private apiToken: string;
  private axios: AxiosInstance;

  constructor(
    baseUrl: string = process.env.NOCODB_URL || '',
    apiToken: string = process.env.NOCODB_TOKEN || ''
  ) {
    if (!baseUrl || !apiToken) {
      throw new Error('NOCODB_URL and NOCODB_TOKEN environment variables are required');
    }

    this.baseUrl = baseUrl.replace(/\/$/, ''); // 末尾スラッシュ削除
    this.apiToken = apiToken;
    this.axios = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'xc-token': apiToken,
        'Content-Type': 'application/json'
      },
      timeout: 10000
    });
  }

  /**
   * レコード一覧取得（Airtable互換）
   */
  async list(baseId: string, tableName: string, options: ListOptions = {}): Promise<AirtableRecord[]> {
    const { where, limit = 100, offset = 0, sort, fields } = options;

    const params: any = { limit, offset };
    if (where) params.where = where;
    if (sort) params.sort = sort;
    if (fields) params.fields = fields.join(',');

    try {
      const projectId = await this.resolveProjectId(baseId);
      const nocodbTableName = this.resolveTableName(baseId, tableName);

      const response = await this.axios.get(
        `/api/v1/db/data/noco/${projectId}/${nocodbTableName}`,
        { params }
      );

      // Airtable形式に変換
      return response.data.list.map((record: any) => ({
        id: String(record.ID || record.Id || record.id),
        fields: this._extractFields(record),
        createdTime: record.CreatedAt || record.created_at || new Date().toISOString()
      }));
    } catch (error: any) {
      if (error.response && error.response.status === 404) {
        const projectId = await this.resolveProjectId(baseId);
        const nocodbTableName = this.resolveTableName(baseId, tableName);
        throw new Error(
          `NocoDB table not found: ${nocodbTableName} (project: ${projectId}, base: ${baseId}, table: ${tableName}). ` +
          `Please verify the table exists in NocoDB.`
        );
      }
      throw new Error(`NocoDB list error: ${error.message}`);
    }
  }

  /**
   * 単一レコード取得
   */
  async get(baseId: string, tableName: string, recordId: string): Promise<AirtableRecord> {
    try {
      const projectId = await this.resolveProjectId(baseId);
      const nocodbTableName = this.resolveTableName(baseId, tableName);

      const response = await this.axios.get(
        `/api/v1/db/data/noco/${projectId}/${nocodbTableName}/${recordId}`
      );

      return {
        id: String(response.data.ID || response.data.Id || response.data.id),
        fields: this._extractFields(response.data),
        createdTime: response.data.CreatedAt || response.data.created_at || new Date().toISOString()
      };
    } catch (error: any) {
      if (error.response && error.response.status === 404) {
        const projectId = await this.resolveProjectId(baseId);
        const nocodbTableName = this.resolveTableName(baseId, tableName);
        throw new Error(
          `NocoDB record not found: ${nocodbTableName}/${recordId} (project: ${projectId}, base: ${baseId}, table: ${tableName})`
        );
      }
      throw new Error(`NocoDB get error: ${error.message}`);
    }
  }

  /**
   * レコード作成
   */
  async create(baseId: string, tableName: string, fields: Record<string, any>): Promise<AirtableRecord> {
    try {
      const projectId = await this.resolveProjectId(baseId);
      const nocodbTableName = this.resolveTableName(baseId, tableName);

      const response = await this.axios.post(
        `/api/v1/db/data/noco/${projectId}/${nocodbTableName}`,
        fields
      );

      return {
        id: String(response.data.ID || response.data.Id || response.data.id),
        fields: this._extractFields(response.data),
        createdTime: response.data.CreatedAt || new Date().toISOString()
      };
    } catch (error: any) {
      const detail = error.response?.data ? JSON.stringify(error.response.data) : error.message;
      throw new Error(`NocoDB create error: ${detail}`);
    }
  }

  /**
   * レコード更新
   */
  async update(baseId: string, tableName: string, recordId: string, fields: Record<string, any>): Promise<AirtableRecord> {
    try {
      const projectId = await this.resolveProjectId(baseId);
      const nocodbTableName = this.resolveTableName(baseId, tableName);

      const response = await this.axios.patch(
        `/api/v1/db/data/noco/${projectId}/${nocodbTableName}/${recordId}`,
        fields
      );

      return {
        id: String(response.data.ID || response.data.Id || response.data.id),
        fields: this._extractFields(response.data),
        createdTime: response.data.CreatedAt || new Date().toISOString()
      };
    } catch (error: any) {
      throw new Error(`NocoDB update error: ${error.message}`);
    }
  }

  /**
   * レコード削除
   */
  async delete(baseId: string, tableName: string, recordId: string): Promise<void> {
    try {
      const projectId = await this.resolveProjectId(baseId);
      const nocodbTableName = this.resolveTableName(baseId, tableName);

      await this.axios.delete(
        `/api/v1/db/data/noco/${projectId}/${nocodbTableName}/${recordId}`
      );
    } catch (error: any) {
      throw new Error(`NocoDB delete error: ${error.message}`);
    }
  }

  // ============================================
  // Meta API (v2) - テーブル・列定義の操作
  // ============================================

  /**
   * テーブルメタデータ取得（列定義含む）
   * @param tableId NocoDB テーブルID（例: mxsy93mwfdvhug1）
   */
  async getTableMeta(tableId: string): Promise<TableMeta> {
    try {
      const response = await this.axios.get(`/api/v2/meta/tables/${tableId}`);

      return {
        id: response.data.id,
        title: response.data.title,
        columns: response.data.columns.map((col: any) => ({
          id: col.id,
          title: col.title,
          uidt: col.uidt,
          dtxp: col.dtxp,
          colOptions: col.colOptions
        }))
      };
    } catch (error: any) {
      const detail = error.response?.data ? JSON.stringify(error.response.data) : error.message;
      throw new Error(`NocoDB getTableMeta error: ${detail}`);
    }
  }

  /**
   * 列オプション更新（SingleSelect/MultiSelect用）
   * @param columnId 列ID（例: ctqf64uyeb7frf9）
   * @param options 更新オプション
   */
  async updateColumn(columnId: string, options: ColumnUpdateOptions): Promise<any> {
    try {
      const response = await this.axios.patch(
        `/api/v2/meta/columns/${columnId}`,
        options
      );
      return response.data;
    } catch (error: any) {
      const detail = error.response?.data ? JSON.stringify(error.response.data) : error.message;
      throw new Error(`NocoDB updateColumn error: ${detail}`);
    }
  }

  /**
   * レコードからフィールドデータを抽出
   */
  private _extractFields(record: any): Record<string, any> {
    const systemFields = ['ID', 'Id', 'id', 'CreatedAt', 'created_at', 'UpdatedAt', 'updated_at'];
    const fields: Record<string, any> = {};

    for (const [key, value] of Object.entries(record)) {
      if (!systemFields.includes(key)) {
        fields[key] = value;
      }
    }

    return fields;
  }

  /**
   * プロジェクトIDをbase_idから解決
   */
  async resolveProjectId(baseId: string): Promise<string> {
    try {
      const mappingPath = join(__dirname, '../nocodb-project-mapping.json');

      if (existsSync(mappingPath)) {
        const mapping = JSON.parse(readFileSync(mappingPath, 'utf-8'));
        return mapping[baseId];
      }

      throw new Error(`NocoDB project mapping not found for base_id: ${baseId}`);
    } catch (error: any) {
      throw new Error(`Failed to resolve project ID: ${error.message}`);
    }
  }

  /**
   * base_idからbase_nameを解決
   */
  resolveBaseName(baseId: string): string {
    try {
      const mappingPath = join(__dirname, '../nocodb-basename-mapping.json');

      if (existsSync(mappingPath)) {
        const mapping = JSON.parse(readFileSync(mappingPath, 'utf-8'));

        if (!mapping.hasOwnProperty(baseId)) {
          throw new Error(`Base name not found in mapping for base_id: ${baseId}`);
        }

        return mapping[baseId];
      }

      throw new Error(`NocoDB basename mapping file not found: ${mappingPath}`);
    } catch (error: any) {
      throw new Error(`Failed to resolve base name: ${error.message}`);
    }
  }

  /**
   * NocoDBテーブル名を構築
   */
  resolveTableName(baseId: string, tableName: string): string {
    const baseName = this.resolveBaseName(baseId);
    const normalizedTableName = tableName.trim();

    if (!normalizedTableName) {
      throw new Error(`Table name cannot be empty for base_id: ${baseId}`);
    }

    if (!baseName || baseName === '') {
      return normalizedTableName;
    }

    return `${baseName}_${normalizedTableName}`;
  }
}
